pkg/controllers/federatedcluster/clusterjoin.go

```diff
@@ -80,232 +75,162 @@ const (
-			); updateErr != nil {
-				return cluster, fmt.Errorf("failed to update cluster status after join timeout: %w", updateErr)
-			}
-
-			return cluster, nil
-		}
-
-		logger.Info("Continue cluster join")
-	}
-
-	// 2. The remaining steps require a cluster kube client, attempt to create one
-
-	restConfig := &rest.Config{Host: cluster.Spec.APIEndpoint}
-
-	clusterSecretName := cluster.Spec.SecretRef.Name
-	if clusterSecretName == "" {
+	if joinedCondition != nil &&
+		joinedCondition.Status == corev1.ConditionFalse &&
+		time.Since(joinedCondition.LastTransitionTime.Time) > clusterJoinTimeout {
+		// join timed out
+		logger.Error(nil, "Cluster join timed out")
 		eventRecorder.Eventf(
 			cluster,
 			corev1.EventTypeWarning,
-			EventReasonJoinClusterError,
-			"cluster %q secret is not set",
-			cluster.Name,
+			EventReasonJoinClusterTimeoutExceeded,
+			"Cluster join timed out",
 		)
-		return cluster, fmt.Errorf("cluster secret is not set")
+		return cluster, &fedcorev1a1.ClusterCondition{
+			Status:  corev1.ConditionFalse,
+			Reason:  JoinTimeoutExceededReason,
+			Message: fmt.Sprintf(JoinTimeoutExceededMessageTemplate, joinedCondition.Message),
+		}, nil
 	}
-	clusterSecret, err := kubeClient.CoreV1().Secrets(fedSystemNamespace).Get(context.TODO(), clusterSecretName, metav1.GetOptions{})
-	if err != nil && apierrors.IsNotFound(err) {
+
+	// 2. The remaining steps require a cluster kube client, attempt to create one
+
+	_, clusterKubeClient, err := getClusterClient(ctx, kubeClient, fedSystemNamespace, cluster)
+	if err != nil {
+		logger.Error(err, "Failed to create cluster client")
+		msg := fmt.Sprintf("Failed to create cluster client: %v", err.Error())
 		eventRecorder.Eventf(
 			cluster,
-			corev1.EventTypeWarning,
-			EventReasonJoinClusterError,
-			"cluster %q secret %q not found",
-			cluster.Name,
-			clusterSecretName,
+			corev1.EventTypeWarning, EventReasonJoinClusterError, msg,
 		)
-		return cluster, fmt.Errorf("cluster secret not found: %w", err)
+		return cluster, &fedcorev1a1.ClusterCondition{
+			Status:  corev1.ConditionFalse,
+			Reason:  TokenNotObtainedReason,
+			Message: msg,
+		}, err
 	}
+
+	// 3. Get or create system namespace in the cluster, this will also tell us if the cluster is unjoinable
+
+	logger.Info(fmt.Sprintf("Create or get system namespace %s in cluster", fedSystemNamespace))
+	memberFedNamespace, err := clusterKubeClient.CoreV1().Namespaces().Get(ctx, fedSystemNamespace, metav1.GetOptions{ResourceVersion: "0"})
 	if err != nil {
-		return cluster, fmt.Errorf("failed to get cluster secret: %w", err)
+		if !apierrors.IsNotFound(err) {
+			msg := fmt.Sprintf("Failed to get namespace: %v", err.Error())
+			logger.Error(err, "Failed to get namespace")
+			eventRecorder.Eventf(
+				cluster,
+				corev1.EventTypeWarning, EventReasonJoinClusterError, msg,
+			)
+			return cluster, &fedcorev1a1.ClusterCondition{
+				Status:  corev1.ConditionFalse,
+				Reason:  TokenNotObtainedReason,
```

Comment:
Wrong reason used here? Should it be `GetOrCreateNamespaceFailed` instead?

Reply:
Oh I noticed that `GetOrCreateNamespaceFailed` was removed... Is there a reason for this?

Reply:
It was a mistake to expose `GetOrCreateNamespace failed` in the first place. The only promise to the users is that we will get the service account toke *somehow*, without any guarantee on *how* exactly it is done. Any errors encountered while obtaining the token will be recorded with a `TokenNotObtained` reason.